Added benchmarks related to Java Strings

lucasccordeiro opened the pull request:
This PR contains some benchmarks related to Java Strings and 
also nondet and assume functions from CProver for SV-COMP

PhilippWendler commented:
Checklist items:

- [ ] programs added to new and [appropriately named](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#directory-structure-and-names) directory
- [ ] license present and [acceptable](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#license) (either in separate file or as comment at beginning of program)
- [ ] [contributed-by](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#origin-description-and-attribution) present (either in README file or as comment at beginning of program)
 
- [ ] programs added to a `.set` file of an existing category, or new sub-category established (if justified)
- [ ] intended property matches the corresponding `.prp` file
- [ ] expected answer in file names according to [convention](https://github.com/sosy-lab/sv-benchmarks/blob/master/CONTRIBUTING.md#properties)

For Java programs, we probably need to discuss a few other things as well, e.g., how to get nondet values etc.

lucasccordeiro commented:
All the comments have been implemented as follows: 

(1) programs added to new and appropriately named directory
Implemented as suggested. I have followed the same structure as described in https://sv-comp.sosy-lab.org/2018/benchmarks.php.

(2) license present and acceptable (either in separate file or as comment at beginning of program)
The license is provided as a separate file named LICENSE.

(3) contributed-by present (either in README file or as comment at beginning of program)
The field contributed-by is available in the README file.

(4) programs added to a .set file of an existing category, or new sub-category established (if justified)
Implemented as suggested.

(5) intended property matches the corresponding .prp file
Implemented as suggested

(6) expected answer in file names according to convention
Implemented as suggested. I have followed the same convention as described in https://sv-comp.sosy-lab.org/2018/benchmarks.php.

peterschrammel commented:
The class providing nondetX() methods is in https://github.com/sosy-lab/sv-benchmarks/pull/472/files#diff-0060ac6ff91a5dc4005233f3ad6770ec

lucasccordeiro commented:
@PhilippWendler: Many thanks for the checklist. Please let us know if you have further suggestions about this PR. We intend to submit additional Java benchmarks once we agree with the proposed structure.

lucasccordeiro commented:
@PhilippWendler: May I ask you whether you have further comments about this PR? We intend to submit additional Java benchmarks once we agree with the proposed structure.

PhilippWendler commented:
I don't have suggestions on what to change, but there are many open questions, which can result in changes. Just the first that come to my mind are:
- How to model nondet functions? In a class or an interface? What names should they have (why prefix them with `__VERIFIER_`)? Should these methods really have a body present? What package name to use (we don't own the namespace `svcomp.org`)? What is the semantics of a nondet object?
- What is the name of the target function? Which class, package etc.? (In your property, you use `__VERIFIER_error`, but the code expects the verifier to check whether `assert`s hold.)
- How is the entry point represented in the property? (You use `Main()` in the property but this does not match the source.)
- Should we actually allow different entry points, or use the convention that they all have the same name (like we do for C programs)?
- Add `.class` files to the repository or not?
- Should the `.set` files list class files or source files? (There exist verifiers that parse Java source).
- How do the answers to these questions change if we have programs that consist of multiple classes, like the ones that already exist in the repository?

These questions and probably more would need to be discussed among the participants who are interested in verifying Java programs, and the competition organizers.

dbeyer commented:
Open questions are not yet addressed.

lucasccordeiro commented:
(1) We model nondet functions using one class named `Verifier`. We add a prefix `__VERIFIER_` to all those nondet functions in order to follow the SV-COMP rules. All those methods have a body and we use the package name `org.svcomp` (we can change it to `svcomp.org` if needed);

(2) We have followed the same standard of other Java benchmarks in the SV-COMP repository to encode the property to be checked via an assertion (e.g., https://github.com/sosy-lab/sv-benchmarks/blob/master/java/MinePump/spec1-5/product1/Actions.java);

(3) In our initial set of benchmarks, we have encoded the property having `Main()` as the entry point. We can further discuss if we need different entry points for the Java benchmarks.

(4) We​ provide pre-compiled `.class` files in the sv-benchmarks alongside with the `.java` files to avoid any ambiguity.

(5) We define list class files in the .set files to avoid any ambiguity.

(6) We intend to provide large Java programs that consist of multiple classes, but we need to agree on a common format to submit them.

lucasccordeiro commented:
@dbeyer and @PhilippWendler: Do you have further issues about this PR? Can we move it forward? 

From the SV-COMP mailing list, we have more teams interested in verifying Java programs (in particular, Evgeny Novikov and Florian Frohn who are not currently members of this group so they don't get the notification).

PhilippWendler commented:
I don't agree with some of the answers.

- Why should a class like `Verifier` has a body? We don't use bodies for the similar C functions as well.
- Neither `org.svcomp` nor `svcomp.org` are namespaces that we should use.
- Using a different way to encode a property would probably be more flexible, but we could probably also allow both kinds of properties. Anyway, the property used in this PR does not match the programs from this PR.
- The use of `Main()` in the property does not match the programs from this PR. Probably we need both a class name and a method name in the property anyway.
- The CPAchecker team is not interested in verifying Java byte code and would not participate in a category where class files are used as verification tasks. I see no reason why the class files are used here, if we actually have the source code. Verifiers that want to use the byte code can still compile the files on their own, but the category should be based on the verification of **Java source code**.

So, there first needs to be an agreement of the scope of such a category, then the details about how the property and other stuff like nondet functions are expressed, and then the programs in this PR and in the repository need to be adjusted accordingly. Those that are interested in such a category would thus need to discuss this.

lucasccordeiro commented:
Hi Philipp,

> I don't agree with some of the answers.
>
>    - Why should a class like Verifier has a body? We don't use bodies for
>    the similar C functions as well.
>
> The C programming language is literally different from Java. What is your
proposal for the Verifier class? Please, make a proposal and then we can
decide which one fits better into SV-COMP.


>
>    - Neither org.svcomp nor svcomp.org are namespaces that we should use.
>
> Which namespace should we actually use?

>
>    - Using a different way to encode a property would probably be more
>    flexible, but we could probably also allow both kinds of properties.
>    Anyway, the property used in this PR does not match the programs from this
>    PR.
>
> May I ask you to be more specific? Which one does not match?

>
>    - The use of Main() in the property does not match the programs from
>    this PR. Probably we need both a class name and a method name in the
>    property anyway.
>
> May I ask you to make a proposal here?

>
>    - The CPAchecker team is not interested in verifying Java byte code
>    and would not participate in a category where class files are used as
>    verification tasks. I see no reason why the class files are used here, if
>    we actually have the source code. Verifiers that want to use the byte code
>    can still compile the files on their own, but the category should be based
>    on the verification of *Java source code*.
>
> What do the other participants think about that? May I ask you to
invite Evgeny Novikov (novikov@ispras.ru) and Florian Frohn (
flofrohn@googlemail.com) to this group so they can participate in this
discussion?

PhilippWendler commented:
> The C programming language is literally different from Java. What is your
> proposal for the Verifier class? Please, make a proposal and then we can
> decide which one fits better into SV-COMP.

Sorry, I don't have time to make a proposal now. Fixing SV-COMP benchmarks is more urgent. (This is just a demo category anyway and does not need to be solved now.) Maybe we can use an interface? Though we would need to get an instance from somewhere... Hmm.

>> - Neither org.svcomp nor svcomp.org are namespaces that we should use.
>
> Which namespace should we actually use?

Not sure, but definitely not a namespace that does not belong to us. Currently I see only `org.sosy_lab.svcomp` as possibility, or somebody would need to register `svcomp.org`.

>> - Using a different way to encode a property would probably be more flexible, but we could probably also allow both kinds of properties. Anyway, the property used in this PR does not match the programs from this PR.
>
> May I ask you to be more specific? Which one does not match?

The property mentions `__VERIFIER_error`, but the programs never call this.

>> - The use of Main() in the property does not match the programs from this PR. Probably we need both a class name and a method name in the property anyway.
>
> May I ask you to make a proposal here?

Sorry, no time currently.

> May I ask you to invite Evgeny Novikov and Florian Frohn to this group so they can participate in this discussion?

@eunovm is already in this group. Anbody can participate in this discussion anyway, but I also sent an invite to Florian.

lucasccordeiro commented:
Fair point. Let's go back to this discussion after fixing the SV-COMP
benchmarks. In the meantime, it would be nice to get the opinion of @eunovm
and Florian on this matter.

On 16 November 2017 at 09:52, Philipp Wendler <notifications@github.com>
wrote:

> The C programming language is literally different from Java. What is your
> proposal for the Verifier class? Please, make a proposal and then we can
> decide which one fits better into SV-COMP.
>
> Sorry, I don't have time to make a proposal now. Fixing SV-COMP benchmarks
> is more urgent. (This is just a demo category anyway and does not need to
> be solved now.) Maybe we can use an interface? Though we would need to get
> an instance from somewhere... Hmm.
>
>
>    - Neither org.svcomp nor svcomp.org are namespaces that we should use.
>
> Which namespace should we actually use?
>
> Not sure, but definitely not a namespace that does not belong to us.
> Currently I see only org.sosy_lab.svcomp as possibility, or somebody
> would need to register svcomp.org.
>
>
>    - Using a different way to encode a property would probably be more
>    flexible, but we could probably also allow both kinds of properties.
>    Anyway, the property used in this PR does not match the programs from this
>    PR.
>
> May I ask you to be more specific? Which one does not match?
>
> The property mentions __VERIFIER_error, but the programs never call this.
>
>
>    - The use of Main() in the property does not match the programs from
>    this PR. Probably we need both a class name and a method name in the
>    property anyway.
>
> May I ask you to make a proposal here?
>
> Sorry, no time currently.
>
> May I ask you to invite Evgeny Novikov and Florian Frohn to this group so
> they can participate in this discussion?
>
> @eunovm <https://github.com/eunovm> is already in this group. Anbody can
> participate in this discussion anyway, but I also sent an invite to Florian.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sosy-lab/sv-benchmarks/pull/472#issuecomment-344871643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADheHWh5QGXLI5jtJ28QLouEF3t7MQibks5s3AXTgaJpZM4PWnQT>
> .
>

peterschrammel commented:
@PhilippWendler, the proposal for the Verifier class is part of the PR: https://github.com/sosy-lab/sv-benchmarks/pull/472/commits/cd3547321f0007960551c0a6f76e7ea8297023df#diff-0060ac6ff91a5dc4005233f3ad6770ec

eunovm commented:
I don't have much experience both with development of Java programs and, especially, with their verification. So, I can't reason about technical issues in details.

From the generic point of view, due to popularity of Java, we can easily assume that having "heavyweight" static verification tools is welcome, especially, for critical parts of programs. First of all you can target generic correctness rules, say, there shouldn't be (unhandled) exceptions (likely different exception should be considered separately), programs should properly use standard libraries and so on. Verification of functional correctness is always too time-consuming, so it should be most likely out of the scope.

Regarding to verification of source/bytecode files, one can have just compiled code but still needs to verify it. If it will be possible, it will be great. As far as I understand, there is much less differences between source files and bytecode files in Java than, say, between source and compiled files in C. But there still can be some substantial ones. From this point of view, abilities in verification of source or bytecode files can be interesting independently. But it is better to hear an opinion of Java experts.

lucasccordeiro commented:
@PhilippWendler: May I ask you whether you had a chance to work on your proposal? It would be nice if we could run the Java demo category for SV-COMP 2018.

PhilippWendler commented:
I am planning on having it ready for demo categories, but nothing won't happen over the holidays now.

peterschrammel commented:
Obsolete due to https://github.com/sosy-lab/sv-benchmarks/pull/565.
